File: ir-core/src/main/kotlin/com/voxeldev/mcodegen/dsl/ir/IrAnnotation.kt
package com.voxeldev.mcodegen.dsl.ir

import com.voxeldev.mcodegen.dsl.ir.utils.MapStringAnySerializer
import kotlinx.serialization.Serializable

/**
 * Represents an annotation in the IR (Intermediate Representation) system.
 * Annotations provide additional metadata about code elements.
 */
@Serializable
data class IrAnnotation(
    val name: String,
    val parameters: List<IrAnnotationParameter>,
    @Serializable(with = MapStringAnySerializer::class)
    val languageProperties: Map<String, Any>,
)

@Serializable
data class IrAnnotationParameter(
    val parameterName: String,
    val parameterValue: IrExpression,
)

File: swift-companion/Sources/IrCore/IrAnnotation.swift
//
//  IrAnnotation.swift
//  swift-companion
//
//  Created by nVoxel on 17.06.2025.
//

import Foundation
import AnyCodable

/**
 * Represents an annotation in the IR (Intermediate Representation) system.
 * Annotations provide additional metadata about code elements.
 */
public struct IrAnnotation: Codable {
    public let type: String?
    public let name: String
    public let parameters: [IrAnnotationParameter]
    /// Arbitrary language-specific key/value pairs.
    public let languageProperties: [String: AnyCodable]

    public init(
        name: String,
        parameters: [IrAnnotationParameter],
        languageProperties: [String: AnyCodable] = [:]
    ) {
        self.type = "com.voxeldev.mcodegen.dsl.ir.IrAnnotation"
        self.name = name
        self.parameters = parameters
        self.languageProperties = languageProperties
    }
}

/**
 * Represents a parameter of an annotation in the IR (Intermediate Representation) system.
 */
public struct IrAnnotationParameter: Codable {
    public let type: String?
    public let parameterName: String
    public let parameterValue: IrExpression

    public init(parameterName: String, parameterValue: IrExpression) {
        self.type = "com.voxeldev.mcodegen.dsl.ir.IrAnnotationParameter"
        self.parameterName = parameterName
        self.parameterValue = parameterValue
    }
}
